replica/table_sstable_set.cc

```diff
@@ -0,0 +1,240 @@
+        return stop_iteration::no;
+    });
+    return ret;
+}
+
+uint64_t
+table_sstable_set::bytes_on_disk() const noexcept {
+    uint64_t ret = 0;
+    for_each_sstable_set_until(query::full_partition_range, [&] (lw_shared_ptr<sstable_set> set) {
+        ret += set->bytes_on_disk();
+        return stop_iteration::no;
+    });
+    return ret;
+}
+
+class table_incremental_selector : public incremental_selector_impl {
+    table& _table;
+
+    // _cur_set and _cur_selector contain a snapshot
+    // for the currently selected compaction_group.
+    // _lowest_next_position corresponds the start position
+    // of the compaction_group following the currently selected one
+    lw_shared_ptr<sstable_set> _cur_set;
+    std::optional<sstable_set::incremental_selector> _cur_selector;
+    dht::ring_position_ext _lowest_next_position = dht::ring_position_ext::max();
+
+public:
+    table_incremental_selector(table& t)
+            : _table(t)
+    {
+        if (auto cg = _table.single_compaction_group_if_available()) {
+            _cur_set = cg->make_compound_sstable_set();
+            _cur_selector.emplace(_cur_set->make_incremental_selector());
+        }
+    }
+
+    virtual std::tuple<dht::partition_range, std::vector<shared_sstable>, dht::ring_position_ext> select(const dht::ring_position_view& pos) override {
+        // Always return minimum singular range, such that incremental_selector::select() will always call this function,
+        // which in turn will find the next sstable set to select sstables from.
+        const dht::partition_range current_range = dht::partition_range::make_singular(dht::ring_position::min());
+
+        if (!_cur_selector) {
+            auto* sg = _table.storage_group_for_token(pos.token());
+            _cur_set = sg->make_compound_sstable_set();
+            _cur_selector.emplace(_cur_set->make_incremental_selector());
+            auto next_token = sg->main_compaction_group()->next_compaction_group_token();
+            _lowest_next_position = next_token.is_maximum()
+                    ? _lowest_next_position = dht::ring_position_ext::max()
+                    : dht::ring_position_ext::starting_at(next_token);
+        }
+
+        auto res = _cur_selector->select(pos);
+        // Return all sstables selected on the requested position from the first matching sstable set.
+        // This assumes that the underlying sstable sets are disjoint in their token ranges so
+        // only one of them contain any given token.
+        auto sstables = std::move(res.sstables);
+        // Return the lowest next position, such that this function will be called again to select the
+        // lowest next position from the selector which previously returned it.
+        // Until the current selector is exhausted. In that case,
+        // jump to the next compaction_group sstable set.
+        dht::ring_position_ext lowest_next_position = res.next_position;
+        if (lowest_next_position.is_max()) {
```

Comment:
When this is true, we are done, no? Nothing more to select?

Reply:
only in the current group we're selecting from.
this is the que to move on to the next group

Reply:
I find overloading the max-token as a cue to move to the next group a bit confusing.